src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,295 @@
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// Represents a Utf8JsonWriter for writing JSON data to a stream.
+    /// </summary>
+    internal sealed partial class ODataUtf8JsonWriter : IJsonWriter, IDisposable, IAsyncDisposable
+    {
+        /// <summary>
+        /// Starts a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A stream for writing the stream value.</returns>
+        public Stream StartStreamValueScope()
+        {
+            this.CommitUtf8JsonWriterContentsToBuffer();
+
+            this.WriteItemWithSeparatorIfNeeded();
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            this.Flush();
+
+            this.binaryValueStream =  new ODataUtf8JsonWriterStream(this);
+
+            return this.binaryValueStream;
+        }
+
+        /// <summary>
+        /// Ends a scope for writing a stream value.
+        /// </summary>
+        public void EndStreamValueScope()
+        {
+            this.binaryValueStream.Flush();
+            this.binaryValueStream.Dispose();
+            this.binaryValueStream = null;
+            this.Flush();
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Asynchronously starts a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation. The task result contains a stream for writing the stream value.</returns>
+        public async Task<Stream> StartStreamValueScopeAsync()
+        {
+            this.WriteSeparatorIfNecessary();
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.binaryValueStream = new ODataUtf8JsonWriterStream(this);
+
+            return this.binaryValueStream;
+        }
+
+        /// <summary>
+        /// Asynchronously ends a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndStreamValueScopeAsync()
+        {
+            await this.binaryValueStream.FlushAsync();
+            await this.binaryValueStream.DisposeAsync();
+            this.binaryValueStream = null;
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Represents a stream for writing UTF-8 JSON data to an ODataUtf8JsonWriter.
+        /// </summary>
+        internal class ODataUtf8JsonWriterStream : Stream

```

Comment:
We return an instance of this stream class from both `StartStreamValueScope` and `StartStreamValueScopeAsync`. What would happens if someone called the `ReadAsync`? Should it not be overridden to return the not supported behaviour? What about methods like `Dispose` and `DisposeAsync`? Should they do any cleanup if they're called.

Reply:
@gathogojr  Is it possible to call ReadAsync when the CanRead property is set to false? I think if they do, they'll automatically get an exception that the particular stream does not support reading.. For the `Begin` and `End` pattern I think this is an old pattern that we shouldn't worry ourselves with supporting and should encourage people to use the Task async pattern.. I don't know what you think here?

Reply:
It's true that `CanRead` property is set to false but why then would it be necessary to override `Read` to throw a `NotSupportedException` but not `ReadAsync`? I'll try and see what happens when someone subclasses `Stream` class, sets `CanRead` property to false and then tries to call `Read` or `ReadAsync`...